internal/eval/store_eval.go

```diff
@@ -689,37 +689,102 @@ func evalGETRANGE(args []string, store *dstore.Store) *EvalResponse {
 	}
 }
 
+// **********************************************************************************************************************************************
+
 // evalZADD adds all the specified members with the specified scores to the sorted set stored at key.
 // If a specified member is already a member of the sorted set, the score is updated and the element
 // reinserted at the right position to ensure the correct ordering.
 // If key does not exist, a new sorted set with the specified members as sole members is created.
 func evalZADD(args []string, store *dstore.Store) *EvalResponse {
-	if len(args) < 3 || len(args)%2 == 0 {
+	// if length of command is 3, throw error as it is not possible
+	if len(args) < 3 {
 		return &EvalResponse{
 			Result: nil,
 			Error:  diceerrors.ErrWrongArgumentCount("ZADD"),
 		}
 	}
-
+	// fetch key and sort the set
 	key := args[0]
-	obj := store.Get(key)
-	var sortedSet *sortedset.Set
+	sortedSet, err := getOrCreateSortedSet(store, key)
+	if err != nil {
+		return &EvalResponse{
+			Result: nil,
+			Error:  err,
+		}
+	}
+	// flags parsing
+	flags, nextIndex := parseFlags(args[1:])
+	if nextIndex >= len(args) || (len(args)-nextIndex)%2 != 0 {
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrWrongArgumentCount("ZADD"),
+		}
+	}
+	// only valid flags works
+	if err := validateFlagsAndArgs(args[nextIndex:], flags); err != nil {
+		return &EvalResponse{
+			Result: nil,
+			Error:  err,
+		}
+	}
+	// all processing takes place here
+	return processMembersWithFlags(args[nextIndex:], sortedSet, store, key, flags)
+}
 
-	if obj != nil {
-		var err []byte
-		sortedSet, err = sortedset.FromObject(obj)
-		if err != nil {
-			return &EvalResponse{
-				Result: nil,
-				Error:  diceerrors.ErrWrongTypeOperation,
-			}
+// parseFlags identifies and parses the flags used in ZADD.
+func parseFlags(args []string) (parsedFlags map[string]bool, nextIndex int) {
+	parsedFlags = map[string]bool{
+		"NX":   false,
+		"XX":   false,
+		"LT":   false,
+		"GT":   false,
+		"CH":   false,
+		"INCR": false,
+	}
+	for i := 0; i < len(args); i++ {
+		switch strings.ToUpper(args[i]) {
+		case "NX":
+			parsedFlags["NX"] = true
+		case "XX":
+			parsedFlags["XX"] = true
+		case "LT":
+			parsedFlags["LT"] = true
+		case "GT":
+			parsedFlags["GT"] = true
+		case "CH":
+			parsedFlags["CH"] = true
+		case "INCR":
+			parsedFlags["INCR"] = true
+		default:
+			return parsedFlags, i + 1
 		}
-	} else {
-		sortedSet = sortedset.New()
 	}
 
-	added := 0
-	for i := 1; i < len(args); i += 2 {
+	return parsedFlags, len(args) + 1
```

Comment:
I'm wondering if this could lead to an off by one error?

When the loop encounters the first non-flag argument (which should be the score), it returns i + 1 as nextIndex.

This causes the actual first non-flag argument to be skipped in subsequent processing because args[nextIndex:] starts from the index after the non-flag argument.

Maybe we can rewrite this as:
```go
var i int
    for i = 0; i < len(args); i++ {
        switch strings.ToUpper(args[i]) {
        case "NX":
            parsedFlags["NX"] = true
        case "XX":
            parsedFlags["XX"] = true
        case "LT":
            parsedFlags["LT"] = true
        case "GT":
            parsedFlags["GT"] = true
        case "CH":
            parsedFlags["CH"] = true
        case "INCR":
            parsedFlags["INCR"] = true
        default:
            // Non-flag argument encountered
            return parsedFlags, i
        }
    }
    return parsedFlags, i
```

Feel free to correct if my understanding is wrong.

Reply:
Hello Jyotinder,

Let me clarify why nextIndex correctly points to the beginning of the score-member pairs without causing any off-by-one errors. Here’s how it works:

**Argument Structure:**

The args array is structured such that the first element (args[0]) is always the key for the sorted set.
The score-member pairs start immediately after the key in args, at args[1].

**Flag Parsing Logic:**

The parseFlags function checks each element in args starting from args[1] to see if it’s a flag (like NX, XX, LT, etc.).
If a flag is detected, parseFlags sets the appropriate flag in parsedFlags and continues parsing.
When it encounters a non-flag (i.e., the first score in the score-member pairs), it returns nextIndex as the position of this first score.

**Position of nextIndex:**

After parsing, nextIndex points to 1, the start of the score-member pairs in args. This ensures that when we call args[nextIndex:], it includes everything from args[1] onward.
Since args[0] is the key and nextIndex = 1, starting from args[1] correctly includes the full sequence of score-member pairs.

I hope this clarifies the logic behind parsing!

Thanks